File: src/CSharp/401-500/437.PathSumIII.cs
using System;
using System.Collections.Generic;

namespace PeefyLeetCode.PathSumIII
{
    public class TreeNode
    {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int x) { val = x; }
    }

public class Solution {
private int count = 0;
    public void dfs(TreeNode p, int sum, int pathSum, Dictionary<int, int> preDict) {
        if (p != null) {
            pathSum += p.val;
            if (preDict.TryGetValue(pathSum - sum, out var temp) == true)
            {
                count += temp;
            }
            if(preDict.ContainsKey(pathSum) == false){
                preDict.Add(pathSum, 1);
            }else{
                preDict[pathSum] = preDict[pathSum] + 1;
            }
            dfs(p.left, sum, pathSum, preDict);
            dfs(p.right, sum, pathSum, preDict);
            preDict[pathSum] = preDict[pathSum] - 1;
        }
    }

    public int PathSum(TreeNode root, int sum) {
        count = 0;
        Dictionary<int, int> preDict = new Dictionary<int, int>();
        preDict.Add(0, 1);
        dfs(root, sum, 0, preDict);
        return count;
    }
}

}

File: src/Python/401-500/437.PathSumIII.py

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def pathSumValue(self, node, sum, val):
        if node is None:
            return 0    
        count = 0     
        val += node.val
        if val == sum:
            count += 1
        return count + self.pathSumValue(node.left, sum, val) + self.pathSumValue(node.right, sum, val)
 
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if root is None:
            return 0 
        return self.pathSum(root.left, sum) + self.pathSum(root.right, sum) + self.pathSumValue(root, sum, 0)

if __name__ == '__main__':
    solution = Solution()
    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(-3)
    root.right.right = TreeNode(11)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(2)
    root.left.left.left = TreeNode(3)
    root.left.left.right = TreeNode(-2)
    root.left.right.right = TreeNode(1)
    print(solution.pathSum(root, 8))
    root = TreeNode(1)
    root.right = TreeNode(2)
    root.right.right = TreeNode(3)
    root.right.right.right = TreeNode(4)
    root.right.right.right.right = TreeNode(5)
    print(solution.pathSum(root, 3))
else:
    pass
